homeassistant/components/hunterdouglas_powerview/__init__.py

```diff
@@ -1 +1,196 @@
+
+    config = entry.data
+
+    hub_address = config.get(CONF_HOST)
+    websession = async_get_clientsession(hass)
+
+    pv_request = AioRequest(hub_address, loop=hass.loop, websession=websession)
+
+    hub = Hub(pv_request)
+
+    try:
+        async with async_timeout.timeout(10):
+            await hub.query_user_data()
+    except HUB_EXCEPTIONS:
+        _LOGGER.error("Connection error to PowerView hub: %s", hub_address)
+        raise ConfigEntryNotReady
+    if not hub.ip:
+        _LOGGER.error("Unable to initialize PowerView hub: %s", hub_address)
+        raise ConfigEntryNotReady
+
+    device_info = await _async_get_device_info(hub, pv_request)
+
+    rooms = Rooms(pv_request)
+    room_data = _async_map_data_by_id((await rooms.get_resources())[ROOM_DATA])
+
+    scenes = Scenes(pv_request)
+    scene_data = _async_map_data_by_id((await scenes.get_resources())[SCENE_DATA])
+
+    shades = Shades(pv_request)
+    shade_data = _async_map_data_by_id((await shades.get_resources())[SHADE_DATA])
+
+    async def async_update_data():
+        """Fetch data from shade endpoint."""
+        shades = hass.data[DOMAIN][entry.entry_id][PV_SHADES]
+        async with async_timeout.timeout(10):
+            shade_entries = await shades.get_resources()
+        if not shade_entries:
+            raise UpdateFailed(f"Failed to fetch new shade data.")
+        return _async_map_data_by_id(shade_entries[SHADE_DATA])
+
+    coordinator = DataUpdateCoordinator(
+        hass,
+        _LOGGER,
+        name="powerview hub",
+        update_method=async_update_data,
+        update_interval=timedelta(seconds=60),
+    )
+
+    hass.data[DOMAIN][entry.entry_id] = {
+        PV_API: pv_request,
+        PV_ROOM_DATA: room_data,
+        PV_SCENE_DATA: scene_data,
+        PV_SHADES: shades,
+        PV_SHADE_DATA: shade_data,
+        COORDINATOR: coordinator,
+        DEVICE_INFO: device_info,
+    }
+
+    for component in PLATFORMS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, component)
+        )
+
+    return True
+
+
+async def _async_get_device_info(hub, pv_request):
+    """Determine device info."""
+    userdata = UserData(pv_request)
+    userdata_data = (await userdata.get_resources())[USER_DATA]
+    main_processor_info = userdata_data[FIRMWARE_IN_USERDATA][
+        MAINPROCESSOR_IN_USERDATA_FIRMWARE
+    ]
+    return {
+        DEVICE_NAME: hub.name,
+        DEVICE_MAC_ADDRESS: userdata_data[MAC_ADDRESS_IN_USERDATA],
+        DEVICE_SERIAL_NUMBER: userdata_data[SERIAL_NUMBER_IN_USERDATA],
+        DEVICE_REVISION: main_processor_info[REVISION_IN_MAINPROCESSOR],
+        DEVICE_FIRMWARE: main_processor_info,
+        DEVICE_MODEL: main_processor_info[MODEL_IN_MAINPROCESSOR],
+    }
+
+
+@callback
+def _async_map_data_by_id(data):
+    """Return a dict with the key being the id for a list of entries."""
+    return {entry[ATTR_ID]: entry for entry in data if ATTR_ID in entry}
```

Comment:
This if-check can cause you to loose data. Wouldn't it be better to blow up so you know you got bad data ?

Reply:
Its overkill. removed the if